File: Neetcode/1DDP/wordbreak.cpp
#include <string>
#include <vector>
using namespace std;

class Solution {
 public:
  bool wordBreak(string s, vector<string>& wordDict) {
    int n = s.size(), m;
    vector<bool> dp(n + 1, false);
    dp[0] = true;
    for (int i = 1; i < n + 1; ++i) {
      for (string word : wordDict) {
        m = word.size();
        if (i >= m) {
          if (dp[i - m] && s.substr(i - m, m) == word) {
            dp[i] = true;
          }
        }
      }
    }
    return dp[n];
  }
};

File: Neetcode/1DDP/wordbreak.py
# https://neetcode.io/problems/word-break
# https://leetcode.com/problems/word-break/

from typing import List


class Solution:
    def wordBreak(self, s: str, wordDict: List[str]) -> bool:
        dp = [False] * (len(s) + 1)
        dp[len(s)] = True
        for i in range(len(s) - 1, -1, -1):
            for w in wordDict:
                if (i + len(w)) <= len(s) and s[i : i + len(w)] == w:
                    dp[i] = dp[i + len(w)]
                if dp[i]:
                    break
        return dp[0]


class MySolution:
    def wordBreak(self, s: str, wordDict: List[str]) -> bool:
        dp = [False] * (len(s) + 1)
        dp[0] = True
        for i in range(1, len(s) + 1):
            for word in wordDict:
                if i >= len(word):
                    if dp[i - len(word)] and s[i - len(word) : i] == word:
                        dp[i] = True
        return dp[len(s)]
